cpp/include/raft/neighbors/cagra_types.hpp

```diff
@@ -183,15 +185,36 @@ struct index : ann::index {
   {
     RAFT_EXPECTS(dataset.extent(0) == knn_graph.extent(0),
                  "Dataset and knn_graph must have equal number of rows");
+    RAFT_LOG_INFO("CAGRA constructor, copying arrays");
     raft::copy(dataset_.data_handle(), dataset.data_handle(), dataset.size(), res.get_stream());
     raft::copy(graph_.data_handle(), knn_graph.data_handle(), knn_graph.size(), res.get_stream());
+    dataset_view_ = dataset_.view();
+    graph_view_   = graph_.view();
     res.sync_stream();
   }
 
+  /** Construct an index from dataset and knn_graph arrays */
+  index(raft::device_resources const& res,
+        raft::distance::DistanceType metric,
+        raft::device_matrix_view<const T, IdxT, row_major> dataset,
+        raft::device_matrix_view<const IdxT, IdxT, row_major> knn_graph)
+    : ann::index(),
+      metric_(metric),
+      dataset_(make_device_matrix<T, IdxT>(res, 0, 0)),
+      graph_(make_device_matrix<IdxT, IdxT>(res, 0, 0))
+  {
+    RAFT_EXPECTS(dataset.extent(0) == knn_graph.extent(0),
+                 "Dataset and knn_graph must have equal number of rows");
+    dataset_view_ = dataset;
+    graph_view_   = knn_graph;
+  }
+
  private:
   raft::distance::DistanceType metric_;
   raft::device_matrix<T, IdxT, row_major> dataset_;
   raft::device_matrix<IdxT, IdxT, row_major> graph_;
+  raft::device_matrix_view<const T, IdxT, row_major> dataset_view_;
```

Comment:
We have some types to help us w/ the `sometimes owning` semantics that I think can be useful here. I'm okay merging this PR in for release the way it is, but I think we should consider using something like the `mdbuffer` or `TemporaryDeviceBuffer` in the future. Though I think the `TemporaryDeviceBuffer` will require some adjustments to make it useful for this specific case (tagging @divyegala).

Reply:
+1 to what @cjnolet mentioned here. I think it will be pretty straightforward to transition to mdbuffer once it's available (hopefully very soon).